src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultRazorComponentSearchEngine.cs

```diff
@@ -0,0 +1,83 @@
+﻿using System;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    internal class DefaultRazorComponentSearchEngine : RazorComponentSearchEngine
+    {
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+
+        public DefaultRazorComponentSearchEngine(ForegroundDispatcher dispatcher)
+        {
+            _foregroundDispatcher = dispatcher ?? throw new ArgumentNullException(nameof(dispatcher));
+        }
+
+        /// <summary>Search for a component in a project based on its tag name and fully qualified name.</summary>
+        /// <remarks>
+        /// This method makes several assumptions about the nature of components. First, it assumes that a component
+        /// a given name `Name` will be located in a file `Name.razor`. Second, it assumes that the namespace the
+        /// component is present in has the same name as the assembly its corresponding tag helper is loaded from.
+        /// Implicitly, this method inherits any assumptions made by TrySplitNamespaceAndType.
+        /// </remarks>
+        /// <param name="project">A project to search for the component in.</param>
+        /// <param name="componentName">The name of the component.</param>
+        /// <param name="fullyQualifiedComponentName">The fully qualified name of the component.</param>
+        /// <param name="cancellationToken">A cancellation token, primarily for the document resolver.</param>
+        /// <returns>A tuple containing the document URI and related resources loaded during search.</returns>
+        public async override Task<Tuple<Uri, DocumentSnapshot, RazorCodeDocument>> TryLocateComponent(ProjectSnapshot project, string componentName, string fullyQualifiedComponentName, CancellationToken cancellationToken)
```

Comment:
Hmm, I'd imagine this API would be more of a:
```C#
public async override Task<bool> TryLocateTagHelperDocumentAsync(TagHelperDescriptor tagHelper, out DocumentSnapshot document)
```

Otherwise the caller needs to always iterate over every project. I'm assuming with this API as it is currently you expect to only call it once for the current documents project based on a component. That can somewhat work; however, the assembly name from the TagHelperDescriptor needs to be taken into account. 

If you make the assumption that TagHelperDescriptor's assembly name roughly equates to the ProjectSnapshot's name (it ends in Name.csproj so you'd have to write some fuzzing logic) you could cover the majority of scenarios and even allow you to navigate outside of the current project with minimal effort

Reply:
I don't think it's as simple as passing just the `tagHelperDescriptor` - afaik, there's no guarantee it's the one with the short tag name helper. Are you suggesting splitting the namespace and type name from the fully qualified `tagHelper.Name`? Also, `async` methods don't allow `ref` parameters, so are we okay with ruling out `async` in the future?

Reply:
Also, to check the name of the `ProjectSnapshot`, are you suggesting using `ProjectSnapshot.FilePath` and checking that it ends with `AssemblyName.csproj`?

Reply:
> Are you suggesting splitting the namespace and type name from the fully qualified `tagHelper.Name`

Yup, you make the same assumption in the provider PR anyhow 😄. Turns out that expectation is a reliable one!



> Also, to check the name of the `ProjectSnapshot`, are you suggesting using `ProjectSnapshot.FilePath` and checking that it ends with `AssemblyName.csproj`?

Yup, it's not ideal but for now it'll allow us to rename a large majority of projects and for others it just wont work which is better then it breaking